### PR TITLE
Allow inline vals with alias to constant type

### PR DESCRIPTION
**Pull request by nicolasstucki**
Allow inline vals with alias to constant type and make them constant fold

### CODE REVIEW

**Comment by plokhotnyuk** [comment]
@nicolasstucki Can `constVal` be used instead of `c.eval` from Scala 2 in compile-time?

I have a couple of use cases:
1) [Expensive calculation of LUT tables that should happen in compile-time](https://github.com/plokhotnyuk/jsoniter-scala/blob/f056571b1c67478531678ddae6418efe62302599/jsoniter-scala-core/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonWriter.scala#L2367-L2391). 
2) [Passing of user defined configuration with "pure" functions that should be evaluated and used in compile-time during macro expansions](https://github.com/plokhotnyuk/jsoniter-scala/blob/f056571b1c67478531678ddae6418efe62302599/jsoniter-scala-macros/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala#L424-L433).

**Comment by nicolasstucki** [comment]
@plokhotnyuk that kind of computation is out of the scope of `inline val`. Inline vals are basically an extension of the current final vals with constant RHS with the difference that the rhs must be a constant and hence is guaranteed to be inlined. What you describe seems to be still in the realm of macros.

**Comment by nicolasstucki** [comment]
@plokhotnyuk you could use the following macro (or something simillar)
```scala
  private final val f64Pow5InvSplit: Array[Long] = ${ evalF64Pow5InvSplit }
 
  // Put this in another file (not necessarily another project)
  // Contents of Expr evaluated at compiletime and result lifted into the code to create that array
  import scala.quoted._
  def evalF64Pow5InvSplit(using QuoteContext): Expr[Array[Long]] = Expr {
    val ss = new Array[Long](582)
    var pow5 = BigInt(1)
    var i = 0
    while (i < 582) {
      val inv = ((BigInt(1) << (pow5.bitLength + 121)) / pow5) + 1
      ss(i) = inv.longValue & 0x3FFFFFFFFFFFFFFFL
      ss(i + 1) = (inv >> 62).longValue & 0x3FFFFFFFFFFFFFFFL
      pow5 *= 5
      i += 2
    }
   ss
  }
```

File: compiler/src/dotty/tools/dotc/ast/TreeInfo.scala [hunk 1]
```diff
@@ -510,7 +510,7 @@ trait TypedTreeInfo extends TreeInfo[Type] { self: Trees.Instance[Type] =>
    */
   def constToLiteral(tree: Tree)(implicit ctx: Context): Tree = {
     val tree1 = ConstFold(tree)
-    tree1.tpe.widenTermRefExpr match {
+    tree1.tpe.widenTermRefExpr.dealias match {
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:513]: For later: maybe we should call `.normalized`, as `dealias` is just one of many possible normalization operations on types.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/ast/TreeInfo.scala:513]: Yes, I have that in #8843. But when I tried to normalize it here it did not work. Maybe I need to deeply dealias and then normalize.

**Pull request merged**